### PR TITLE
Add walkReferenceChain JITServer Support

### PR DESCRIPTION
**Pull request by harryyu1994**
The purpose of this change is to avoid using object addresses on JITServer. Using object addresses on JITServer creates problems when GC moves the object (thus changing the addresses). Instead, JITServer must use some kind of handle (a GlobalReference or some other GC root) to hold the object so that JITServer always has the valid object addresses when it tries to perform operations on it.

`walkReferenceChain()` is special in that it needs half of its information from JITServer(TR::Node) and half of its information from JITClient(object address). During the walk, we need vmaccess so that GC doesn't move the object of interest. We can only have vmaccess on JITClient, which means we need to do the walk on JITClient.

This change makes JITServer generate the offset information JITClient needs for the walk and passes to JITClient in the form of a vector.

Originally we have a LocalGCCounter hack where we can detect if the object has been moved to a different location. And when object has been moved, JITClient will abort the compilation. However, this does not work for all gcpolicies and also is not very efficient because JITClient can abort compilations that might not need to be aborted. This change offers a better solution: since the walk is done on JITClient and JITClient has control over vmaccess, object is guaranteed to not be moved during the operation.

I have also noticed a few misuses of globalreference read and vmaccess and have corrected them. We must have vmaccess before we perform a read.

Issue: #8109

Signed-off-by: Harry Yu <harryyu1994@gmail.com>

### CODE REVIEW

**Comment by mpirvu** [comment]
A generic remark: now that we've 'released` a version of JITServer, any change to the communication protocol including adding/removing of messages or APIs needs to increment the version number

**Comment by harryyu1994** [comment]
Incremented MINOR_NUMBER from 1 to 2.

**Comment by harryyu1994** [comment]
Added another change to handle 3 invokeFilterArgumentsWithCombinerHandle methods, they appeared to be without JITServer support. The latest commit can be reviewed separately. I put it in the same PR because 1. they both modify the same set of files 2. both are related to JSR292 3. both added new messageTypes so it's convenient to have them in the same PR.

File: runtime/compiler/control/JITClientCompilationThread.cpp [hunk 1]
```diff
@@ -2373,17 +2389,13 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
...
-            auto comp = compInfoPT->getCompilation();
-            comp->failCompilation<TR::CompilationInterrupted>("Compilation interrupted due to GC");
-            }
-
-         uintptrj_t methodHandle = std::get<0>(client->getRecvData<uintptrj_t>());
-         TR::VMAccessCriticalSection invokeILGenMacros(fe);
+         auto recv = client->getRecvData<uintptrj_t*, std::vector<uintptrj_t> >();
+         TR::VMAccessCriticalSection invokeILGenMacrosParameterCount(fe);
+         uintptrj_t receiverHandle = *std::get<0>(recv);
+         std::vector<uintptrj_t> listOfOffsets = std::get<1>(recv);
```
Comment (mpirvu) [inline runtime/compiler/control/JITClientCompilationThread.cpp]: Again, let's use a const reference here

File: runtime/compiler/control/JITClientCompilationThread.cpp [hunk 2]
```diff
@@ -2392,7 +2404,29 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
          client->write(response, parameterCount);
          }
          break;
-
+      case MessageType::runFEMacro_invokeILGenMacrosArrayLength:
+         {
+         auto recv = client->getRecvData<uintptrj_t*, std::vector<uintptrj_t> >();
+         TR::VMAccessCriticalSection invokeILGenMacrosArrayLength(fe);
+         uintptrj_t receiverHandle = *std::get<0>(recv);
+         std::vector<uintptrj_t> listOfOffsets = std::get<1>(recv);
```
Comment (mpirvu) [inline runtime/compiler/control/JITClientCompilationThread.cpp]: Again, let's use a const reference here

File: runtime/compiler/control/JITClientCompilationThread.cpp [hunk 3]
```diff
@@ -2392,7 +2404,29 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
...
+         client->write(response, arrayLength);
+         }
+         break;
+      case MessageType::runFEMacro_invokeILGenMacrosGetField:
+         {
+         auto recv = client->getRecvData<uintptrj_t*, uintptrj_t, std::vector<uintptrj_t> >();
+         TR::VMAccessCriticalSection invokeILGenMacrosGetField(fe);
+         uintptrj_t receiverHandle = *std::get<0>(recv);
+         uintptrj_t fieldOffset = std::get<1>(recv);
+         std::vector<uintptrj_t> listOfOffsets = std::get<2>(recv);
```
Comment (mpirvu) [inline runtime/compiler/control/JITClientCompilationThread.cpp]: Again, let's use a const reference here

File: runtime/compiler/control/JITClientCompilationThread.cpp [hunk 4]
```diff
@@ -656,13 +656,6 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
...
-         if (compInfoPT->getLastLocalGCCounter() != compInfoPT->getCompilationInfo()->getLocalGCCounter())
-            {
-            // GC happened, fail compilation
-            auto comp = compInfoPT->getCompilation();
-            comp->failCompilation<TR::CompilationInterrupted>("Compilation interrupted due to GC");
-            }
-
          auto recv = client->getRecvData<uintptrj_t, uintptrj_t>();
          uintptrj_t objectPointer = std::get<0>(recv);
          uintptrj_t fieldOffset = std::get<1>(recv);
```
Comment (mpirvu) [inline runtime/compiler/control/JITClientCompilationThread.cpp]: I believe that getIn32FieldAt requires VM access in hand. same for getInt64FieldAt below

File: runtime/compiler/control/JITClientCompilationThread.cpp [hunk 5]
```diff
@@ -699,13 +692,6 @@ handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe, JITServer::Mes
...
          {
-         if (compInfoPT->getLastLocalGCCounter() != compInfoPT->getCompilationInfo()->getLocalGCCounter())
-            {
-            // GC happened, fail compilation
-            auto comp = compInfoPT->getCompilation();
-            comp->failCompilation<TR::CompilationInterrupted>("Compilation interrupted due to GC");
-            }
-
          uintptrj_t objectPointer = std::get<0>(client->getRecvData<uintptrj_t>());
          client->write(response, fe->getArrayLengthInElements(objectPointer));
```
Comment (mpirvu) [inline runtime/compiler/control/JITClientCompilationThread.cpp]: `fe->getArrayLengthInElements` assumes VM access is already held, but I don't think the client will have it here, so we need to acquire is briefly

**Comment by mpirvu** [comment]
Jenkins test sanity xlinuxjit jdk11

**Comment by mpirvu** [comment]
Jenkins test sanity xlinuxjit jdk11

**Comment by mpirvu** [comment]
Jenkins test sanity plinuxjit jdk11

**Comment by mpirvu** [comment]
All tests have passed, hence merging.

**Pull request merged**